File: static/scripts/roulette.js
// Retrieves the spin button and the roulette wheel image elements from the DOM.
let spinButton = document.getElementById("spin");
let wheel = document.getElementById("wheel_img");

// Maps roulette numbers to their corresponding positions on the wheel image.
const numberToPositionIndex = [0, 14, 32, 2, 34, 18, 27, 6, 21, 10, 19, 23, 4, 25, 12, 36, 16, 30, 8, 35, 13, 33, 9, 20, 17, 31, 1, 26, 5, 7, 22, 11, 37, 15, 29, 3, 24, 28]

/**
 * Initiates a spin on the roulette wheel.
 * Disables the spin button to prevent multiple requests, checks the user's balance,
 * and sends a POST request to the server with the bet details.
 * Upon receiving the result, it animates the wheel to the result position.
 */
async function spin() {
    spinButton.disabled = true;
    updateUserdata();
    if (parseInt(document.getElementById("balance").innerText) < parseInt(document.getElementById("bet").value)) {
        alert("You don't have enough balance to make this bet.");
        spinButton.disabled = false;
    } else if (document.getElementById("bet").value < 1) {
        alert("Bet must be greater than 0.");
        spinButton.disabled = false;
    } else {
        wheel.classList.add("spinning_wait");
        wheel.addEventListener("animationend", restartSpinningWait);
        const timeoutPromise = new Promise(resolve => setTimeout(resolve, 2000));
        const fetchPromise = fetch("/roulette/spin", {
            method: "POST",
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify({
                "bet": document.getElementById("bet").value,
                "type": document.getElementById("bet-type").value,
                "choice": document.querySelector('input[name="choice"]:not([style*="display: none"]):checked') ? document.querySelector('input[name="choice"]:checked').value : null,
                "number": document.getElementById("number").style.display !== "none" ? document.getElementById("number").value : null
            })
            // "choice": choice
        }).then(response => response.text());

        Promise.all([fetchPromise, timeoutPromise]).then(async (values) => {
            const result = parseInt(values[0]);
            if (isNaN(result)) {
                alert(values[0]);
                spinButton.disabled = false;
            } else {
                const positionIndex = numberToPositionIndex[result];
                const targetDeg = positionIndex * 9.473684210526316;
                wheel.style.setProperty("--target-deg", `${targetDeg + (2 * 360)}deg`);
                wheel.removeEventListener("animationend", restartSpinningWait);
                wheel.addEventListener("animationend", function () {
                    wheel.classList.remove("spinning_wait");
                    void wheel.offsetWidth;
                    wheel.classList.add("spinning");
                    wheel.addEventListener("animationend", function () {
                        updateUserdata();
                        wheel.style.transform = `rotate(${targetDeg}deg)`;
                        wheel.classList.remove("spinning");
                        wheel.style.setProperty("--start-deg", `${targetDeg}deg`);
                        wheel.style.setProperty("--target-wait-deg", `${targetDeg + 360}deg`);
                        spinButton.disabled = false;
                    }, {once: true});
                }, {once: true});
            }
        });
    }
}

// Adds an event listener to the spin button to initiate the spin process.
spinButton.addEventListener('click', spin);

/**
 * Restart the spinning animation.
 * This function is called after the spinning animation ends.
 */
function restartSpinningWait() {
    wheel.classList.remove("spinning_wait");
    void wheel.offsetWidth;
    wheel.classList.add("spinning_wait");
}

/**
 * Updates the input container based on the selected bet type.
 * This function dynamically changes the input fields to match the bet type selected by the user.
 * It supports three bet types: color, even/odd, and number. Based on the selection,
 * it displays the relevant input fields and hides the others.
 *
 * For the 'color' bet type, it checks the 'red' radio button by default and clears the number input.
 * For the 'evenodd' bet type, it checks the 'even' radio button by default and clears the number input.
 * For the 'number' bet type, it sets the number input's value to 10 and unchecks all choice radio buttons.
 */
function updateInputContainer() {
    const selectedType = document.getElementById('bet-type').value;
    const colorInputs = document.getElementById('colorInputs');
    const evenOddInputs = document.getElementById('evenOddInputs');
    const numberInput = document.getElementById('numberInput');

    // Initially hide all elements
    colorInputs.style.display = 'none';
    evenOddInputs.style.display = 'none';
    numberInput.style.display = 'none';

    // Display the appropriate element based on the selected bet type
    if (selectedType === 'color') {
        document.getElementById('red').checked = true;
        colorInputs.style.display = 'block';
        document.getElementById('number').value = null;
    } else if (selectedType === 'evenodd') {
        document.getElementById('even').checked = true;
        evenOddInputs.style.display = 'block';
        document.getElementById('number').value = null;
    } else if (selectedType === 'number') {
        document.getElementById('number').value = 10;
        numberInput.style.display = 'block';
        Array.from(document.getElementsByClassName('choice-radio-button')).forEach(radio => {
            radio.checked = false;
        })
    }
}

// Initializes the input container and wheel properties when the document is loaded.
document.addEventListener('DOMContentLoaded', updateInputContainer);
document.addEventListener('DOMContentLoaded', function () {
    wheel.style.setProperty("--start-deg", "0deg");
    wheel.style.setProperty("--target-wait-deg", "360deg");
})

// Updates the input container when the bet type is changed.
document.getElementById('bet-type').addEventListener('change', updateInputContainer);

File: blueprints/roulette.py
import random

from flask import Blueprint, request, render_template

from helper_functions.user_administration import userid_from_token, userdata_from_id, update_balance
from helper_functions.stats import played_game

roulette_blueprint = Blueprint('roulette', __name__)


@roulette_blueprint.route('/')
def roulette_home():
    if request.method == "GET":
        return render_template("roulette/roulette.html")


@roulette_blueprint.route('/spin', methods=["POST"])
def spin():
    content = request.json

    user_id = userid_from_token(request.cookies.get('token'))
    user_data = userdata_from_id(user_id)

    if int(user_data['balance']) >= int(content['bet']):
        result = random.randint(0, 37)
        if content['type'] == "number":
            if int(content['number']) is None:
                return "invalid number"
            elif int(content['number']) == result:
                bet_value = 36 * int(content['bet'])
            else:
                bet_value = -int(content['bet'])
        elif content['choice'] is None:
            return "invalid choice"
        elif content['type'] == "evenodd":
            even = content['choice'] == "even" and result % 2 == 0
            odd = content['choice'] == "odd" and result % 2 == 1
            if (even or odd) and result != 0:
                bet_value = int(content['bet'])
            else:
                bet_value = -int(content['bet'])
        elif content['type'] == "color":
            red = [1, 2, 3, 4, 5, 7, 9, 12, 14, 15, 16, 17, 18, 23, 27, 30, 36, 37]
            black = [6, 8, 10, 11, 13, 19, 20, 21, 22, 24, 25, 26, 28, 29, 31, 32, 33, 34, 35]
            b_red = content['choice'] == "red" and result in red
            b_black = content['choice'] == "black" and result in black
            if b_red or b_black:
                bet_value = int(content['bet'])
            else:
                bet_value = -int(content['bet'])
        else:
            return "invalid choice"

        update_balance(user_id, bet_value)
        played_game(user_id, bet_value, "roulette", double_field=result)

        return str(result)
    else:
        return "not enough money"
